scene/3d/chain_ik_3d.cpp

```diff
@@ -0,0 +1,484 @@
+	const LocalVector<BoneJoint> &joints = chain_settings[p_index]->joints;
+	ERR_FAIL_INDEX_V(p_joint, (int)joints.size(), -1);
+	return joints[p_joint].bone;
+}
+
+void ChainIK3D::set_joint_count(int p_index, int p_count) {
+	ERR_FAIL_INDEX(p_index, (int)settings.size());
+	ERR_FAIL_COND(p_count < 0);
+	LocalVector<BoneJoint> &joints = chain_settings[p_index]->joints;
+	joints.resize(p_count);
+	_set_joint_count(p_index, p_count);
+	notify_property_list_changed();
+}
+
+void ChainIK3D::_set_joint_count(int p_index, int p_count) {
+	//
+}
+
+int ChainIK3D::get_joint_count(int p_index) const {
+	ERR_FAIL_INDEX_V(p_index, (int)settings.size(), 0);
+	const LocalVector<BoneJoint> &joints = chain_settings[p_index]->joints;
+	return joints.size();
+}
+
+void ChainIK3D::_bind_methods() {
+	// Setting.
+	ClassDB::bind_method(D_METHOD("set_root_bone_name", "index", "bone_name"), &ChainIK3D::set_root_bone_name);
+	ClassDB::bind_method(D_METHOD("get_root_bone_name", "index"), &ChainIK3D::get_root_bone_name);
+	ClassDB::bind_method(D_METHOD("set_root_bone", "index", "bone"), &ChainIK3D::set_root_bone);
+	ClassDB::bind_method(D_METHOD("get_root_bone", "index"), &ChainIK3D::get_root_bone);
+
+	ClassDB::bind_method(D_METHOD("set_end_bone_name", "index", "bone_name"), &ChainIK3D::set_end_bone_name);
+	ClassDB::bind_method(D_METHOD("get_end_bone_name", "index"), &ChainIK3D::get_end_bone_name);
+	ClassDB::bind_method(D_METHOD("set_end_bone", "index", "bone"), &ChainIK3D::set_end_bone);
+	ClassDB::bind_method(D_METHOD("get_end_bone", "index"), &ChainIK3D::get_end_bone);
+
+	ClassDB::bind_method(D_METHOD("set_extend_end_bone", "index", "enabled"), &ChainIK3D::set_extend_end_bone);
+	ClassDB::bind_method(D_METHOD("is_end_bone_extended", "index"), &ChainIK3D::is_end_bone_extended);
+	ClassDB::bind_method(D_METHOD("set_end_bone_direction", "index", "bone_direction"), &ChainIK3D::set_end_bone_direction);
+	ClassDB::bind_method(D_METHOD("get_end_bone_direction", "index"), &ChainIK3D::get_end_bone_direction);
+	ClassDB::bind_method(D_METHOD("set_end_bone_length", "index", "length"), &ChainIK3D::set_end_bone_length);
+	ClassDB::bind_method(D_METHOD("get_end_bone_length", "index"), &ChainIK3D::get_end_bone_length);
+
+	// Individual joints.
+	ClassDB::bind_method(D_METHOD("get_joint_bone_name", "index", "joint"), &ChainIK3D::get_joint_bone_name);
+	ClassDB::bind_method(D_METHOD("get_joint_bone", "index", "joint"), &ChainIK3D::get_joint_bone);
+
+	ClassDB::bind_method(D_METHOD("get_joint_count", "index"), &ChainIK3D::get_joint_count);
+}
+
+void ChainIK3D::_validate_bone_names() {
+	for (uint32_t i = 0; i < settings.size(); i++) {
+		// Prior bone name.
+		if (!chain_settings[i]->root_bone.name.is_empty()) {
+			set_root_bone_name(i, chain_settings[i]->root_bone.name);
+		} else if (chain_settings[i]->root_bone.bone != -1) {
+			set_root_bone(i, chain_settings[i]->root_bone.bone);
+		}
+		// Prior bone name.
+		if (!chain_settings[i]->end_bone.name.is_empty()) {
```

Comment:
I think this should only run if `extend_end_bone` (since the property is undefined if not `extend_end_bone`)

Reply:
Same as https://github.com/godotengine/godot/pull/110120#discussion_r2479446302